### PR TITLE
Integration of the incompressible Navier-Stokes solver of adaflo into MeltPoolDG

### PR DESCRIPTION
**Pull request by mschreter**
With this PR we will provide a wrapper interface to the fluid solver package adaflo. This will be coupled to simulate two-phase flow problems with level set modules/phase field modules.

### CODE REVIEW

**Comment by mschreter** [comment]
This is the result of today's peer-programming session with Peter.

image

File: include/meltpooldg/flow/adaflo_wrapper.hpp [hunk 1]
```diff
@@ -5,30 +5,153 @@
...
+    if (fluctuating)
+      values(0) *= std::sin(this->get_time()*numbers::PI/8.);
+    for (unsigned int d=1; d<dim; ++d)
+      values(d) = 0;
+  }
+
+  private:
+    const bool fluctuating;
+  };
+
```
Comment (peterrum) [inline include/meltpooldg/flow/adaflo_wrapper.hpp:47]: BC and IC shouldn't be part of the wrapper.

File: include/meltpooldg/flow/adaflo_wrapper.hpp [hunk 2]
```diff
@@ -5,30 +5,153 @@
...
-            FlowParameters params;
+        // set boundary conditions
+        navier_stokes.set_no_slip_boundary(0);
+        navier_stokes.set_velocity_dirichlet_boundary(1, std::shared_ptr<Function<dim> >(new InflowVelocity<dim>(0., true)));
 
-            parallel::distributed::Triangulation<dim> tria(MPI_COMM_WORLD);
+        navier_stokes.set_open_boundary_with_normal_flux(2, std::shared_ptr<Function<dim> > (new Functions::ZeroFunction<dim>(1)));
+
+        // set initial condition
+        navier_stokes.setup_problem(InflowVelocity<dim>(0., false));
```
Comment (peterrum) [inline include/meltpooldg/flow/adaflo_wrapper.hpp:70]: We need to think about a way how to propagate BC/IC to here.

File: include/meltpooldg/flow/adaflo_wrapper_parameters.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// namespace MeltPoolDG
+// {
+//   namespace Flow
+//   {
+//     struct
+//     AdafloWrapperParameters
+//     {
+//         template<typename number>
+//         AdafloWrapperParameters(const Parameters<number>& mp_params)
+//         {
```
Comment (peterrum) [inline include/meltpooldg/flow/adaflo_wrapper_parameters.hpp:19]: I still think that it is the best approach to have parameter structs in separate files to reduce the dependencies of the modules.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    public:
+      TwoPhaseFlowProblem() = default;
+
+      void
+      run(std::shared_ptr<SimulationBase<dim>> base_in) final
+      {
+        initialize(base_in);
+
+        adaflo = std::make_shared<AdafloWrapper<dim>>(*scratch_data, 
+                                                      base_in->parameters.adaflo_params);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:61]: This should be part of initialize?

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+        // output_results(0,base_in->parameters);
+        while (!time_iterator.is_finished())
+        {
+            const double dt = time_iterator.get_next_time_increment();
+            
+            adaflo->solve();
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:68]: Once we an interface class we don't need to call adaflo directly.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+            
+            to_block_vector(adaflo->get_velocity(), advection_velocity);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:70]: Conversion should happen in the wrapper.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+            output_results(time_iterator.get_current_time_step_number(), 
+                            base_in->parameters);
+
+            level_set_operation.solve(dt, advection_velocity);
+
+            BlockVectorType                     surface_tension_force;
+            scratch_data->initialize_dof_vector(surface_tension_force, 0);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:77]: I guess we can move this out of the while loop.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+            level_set_operation.compute_surface_tension(surface_tension_force, 
+                                                        base_in->parameters.flow.surface_tension_coefficient);
+
+            VectorType surface_out;
+            scratch_data->initialize_dof_vector(surface_out, dof_adaflo_idx);
+            from_block_vector(surface_tension_force, surface_out);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:83]: Part of wrapper.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+            adaflo->set_surface_tension(surface_out);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:84]: The resulting algorithm would look like this:
```cpp
        while (!time_iterator.is_finished())
          {
            const double dt = time_iterator.get_next_time_increment();
            flow_operation->solve(dt /*TODO*/);
            level_set_operation->solve(dt, advection_velocity);
            flow->update_phases(level_set_operation->get_phase_indicators()); // TODO
            level_set_operation->compute_surface_tension(surface_tension_force, 
                                                         surface_tension_coefficient);
            flow_operation->set_surface_tension(surface_out);
          }
```
We also need to be able to synchronize the timpestepping schemes (ours vs. that from adaflo).

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        /*
+         *  setup DoFHandler adaflo
+         */
+        dof_handler_adaflo.initialize(*base_in->triangulation,
+                                FESystem<dim>(FE_Q<dim>(base_in->parameters.base.degree), dim));
+        /*
+         *  setup DoFHandler adaflo
+         */
+        dof_handler.initialize(*base_in->triangulation,
+                                FE_Q<dim>(base_in->parameters.base.degree));
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:119]: This should not be here. Maybe we could attach it inside of the wrappers.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+        // scratch_data->attach_dof_handler(dof_handler_adaflo);
+        scratch_data->attach_dof_handler(dof_handler);
+        scratch_data->attach_dof_handler(dof_handler);
+        scratch_data->attach_dof_handler(dof_handler_adaflo);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:124]: ... same here.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+                                                     constraints_dirichlet);
+          }
+        constraints_dirichlet.close();
+
+        // scratch_data->attach_constraint_matrix(dummy_constraint);
+        const unsigned int dof_no_bc_idx =
+          scratch_data->attach_constraint_matrix(hanging_node_constraints);
+        const unsigned int dof_idx        = scratch_data->attach_constraint_matrix(constraints_dirichlet);
+        
+        dof_adaflo_idx = scratch_data->attach_constraint_matrix(dummy_constraints);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:156]: same here.

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+        /*
+         *  create the matrix-free object
+         */
+        scratch_data->build();
+
+        time_iterator.initialize(TimeIteratorData<double>{0.0 /*start*/,
+                                                    8 /*end*/,
+                                                    0.02 /*dt*/,
+                                                    1000 /*max_steps*/,
+                                                    false /*cfl_condition-->not supported yet*/});
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:170]: Parameters :)

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+                 cell.get_dof_values(in.block(d), local_component);
+
+                 for(unsigned int c = 0; c < n_dofs_per_component; ++c)
+                   local[c * dim + d] = local_component[c];
+              }
+              cell_adaflo->set_dof_values(local, out);
+          }
+
+        out.update_ghost_values();
+      }
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:257]: Free functions. As discussed, we could name them `fe_systems_to_block_vector` and `block_to_fe_systems_vector`

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+         *  output advected field
+        */
+        DataOut<dim> data_out;
+        data_out.attach_dof_handler(dof_handler);
+
+        // data_out.add_data_vector(adaflo->get_velocity(), "velocity");
+        for (auto d = 0; d < dim; ++d)
+          data_out.add_data_vector(dof_handler,
+                                   advection_velocity.block(d),
+                                  "advection_velocity_" + std::to_string(d));
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:277]: We might want to query more quantities from the flow solver (pressure, ...).

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+
+        data_out.add_data_vector(level_set_operation.solution_level_set, "level_set");
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:279]: Let's not make these public fields `solution_level_set` ;)

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+
+      BlockVectorType                     advection_velocity;
+
+      unsigned int dof_adaflo_idx;
+      std::shared_ptr<ScratchData<dim>>   scratch_data;
+      std::shared_ptr<AdafloWrapper<dim>> adaflo;
+      LevelSet::LevelSetOperation<dim>    level_set_operation;
+    };
+  } // namespace TwoPhaseFlow
+} // namespace MeltPoolDG
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:316]: New line.

File: include/meltpooldg/interface/parameters.hpp [hunk 18]
```diff
@@ -115,6 +125,29 @@ namespace MeltPoolDG
...
+        params.declare_parameters(prm_adaflo);
+        params.parse_parameters(parameter_filename, prm_adaflo);
+      }
+
+      const FlowParameters& get_parameters() const
+      {
+        return this->params;
+      }
+      FlowParameters params;
+  };
```
Comment (peterrum) [inline include/meltpooldg/interface/parameters.hpp:148]: As stated above, lets move these to new files.

File: include/meltpooldg/interface/problem_selector.hpp [hunk 19]
```diff
@@ -18,15 +19,17 @@ namespace MeltPoolDG
...
 
-      else if (problem_name == "reinitialization")
-        return std::make_shared<Reinitialization::ReinitializationProblem<dim>>();
+      // else if (problem_name == "reinitialization")
+      //   return std::make_shared<Reinitialization::ReinitializationProblem<dim>>();
 
-      else if (problem_name == "advection_diffusion")
-        return std::make_shared<AdvectionDiffusion::AdvectionDiffusionProblem<dim>>();
+      // else if (problem_name == "advection_diffusion")
+      //   return std::make_shared<AdvectionDiffusion::AdvectionDiffusionProblem<dim>>();
```
Comment (peterrum) [inline include/meltpooldg/interface/problem_selector.hpp:29]: ;)

File: include/meltpooldg/interface/simulationbase.hpp [hunk 20]
```diff
@@ -56,14 +57,15 @@ namespace MeltPoolDG
...
-          "this->field_conditions.initial_field = std::make_shared<MyInitializeFunc<dim>>();"));
-    };
+    // @todo: shift to problems
+    //   AssertThrow(
+    //     this->field_conditions.initial_field,
+    //     ExcMessage(
+    //       "It seems that your SimulationBase object does not contain "
+    //       "a valid initial field function. A shared_ptr to your initial field "
+    //       "function, e.g., MyInitializeFunc<dim> must be specified as follows: "
+    //       "this->field_conditions.initial_field = std::make_shared<MyInitializeFunc<dim>>();"));
```
Comment (peterrum) [inline include/meltpooldg/interface/simulationbase.hpp:67]: ;)

File: include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.hpp [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+#include <deal.II/grid/manifold_lib.h>
+#include <iostream>
+// MeltPoolDG
+#include <meltpooldg/interface/simulationbase.hpp>
+
+namespace MeltPoolDG
+{
+  namespace Simulation
+  {
+    namespace FlowPastZylinder
```
Comment (peterrum) [inline include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.hpp:17]: Zylinder ;) Maybe we should add a comment like we have in deal.II realted to `QTrapez`:

```
/**
 * An alias for QTrapezoid available for historic reasons. This name is
 * deprecated.
 *
 * The class was originally named QTrapez, a poorly named choice since the
 * proper name of the quadrature formula
 * is "trapezoidal rule", or sometimes also called the "trapezoid rule". The
 * misnomer resulted from the fact that its original authors' poor English
 * language skills led them to translate the name incorrectly from the German
 * "Trapezregel".
 */

:P
```

File: include/meltpooldg/simulations/simulation_selector.hpp [hunk 22]
```diff
@@ -22,22 +23,27 @@ namespace MeltPoolDG
...
-                                                                            mpi_communicator);
-          }
-        else if (simulation_name == "advection_diffusion")
-          {
-            return std::make_shared<AdvectionDiffusion::SimulationAdvec<dim>>(parameter_file,
-                                                                              mpi_communicator);
-          }
-        else if (simulation_name == "rotating_bubble")
-          {
-            return std::make_shared<RotatingBubble::SimulationRotatingBubble<dim>>(
```
Comment (peterrum) [inline include/meltpooldg/simulations/simulation_selector.hpp:40]: If this is really necessary than we should start to precompile stuff!

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 23]
```diff
@@ -0,0 +1,316 @@
...
+        scratch_data = std::make_shared<ScratchData<dim>>(/*do_matrix_free*/ true);
+        /*
+         *  setup mapping
+         */
+        scratch_data->set_mapping(MappingQGeneric<dim>(base_in->parameters.base.degree));
+        /*
+         *  setup DoFHandler adaflo
+         */
+        dof_handler_adaflo.initialize(*base_in->triangulation,
+                                FESystem<dim>(FE_Q<dim>(base_in->parameters.base.degree), dim));
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:114]: We don't need to create a new DoFHandler; it is already part of `NavierStokes` of adaflo. We could access it directly in the wrapper.

**Review by peterrum — COMMENTED** [review]
> today's peer-programming session

That was fun! Hope it was not the last time!

I have quickly went through our code from yesterday. I have collected some points we need to address. In particular
- [ ] introduce a interface class `FlowBase` (?)
- [ ] update the phases in the flow
- [ ] synchronize the time stepping schemes
- [ ] clean up the `AdafloWrapper` and move IC and BC related stuff
- [ ] move stuff from the problem class to `AdaflowWrapper` (e.g. the way vectors are stored in adflow is specific to this solver)
- [ ] get `rising bubble` running
- [ ] better understand `adaflo`: its features and limitations

I guess I have forgotten some important points. 

I will create a new CI instance with adaflo enabled.

File: include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.json [hunk 24]
```diff
@@ -33,17 +33,17 @@
...
+      "linearization scheme"         : "coupled implicit Newton",
+      "NL max iterations"            : "10",
+      "NL tolerance"                 : "1.e-9",
+      "lin max iterations"           : "30",
+      "lin tolerance"                : "1.e-5",
+      "lin relative tolerance"       : "1",
+      "lin velocity preconditioner"  : "ilu scalar",
+      "lin pressure mass preconditioner" : "ilu",
+      "lin its before inner solvers" : "30"
+    }
```
Comment (peterrum) [inline include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.json]: @mschreter Fixed :) Now ILUs is actually picked up!

Comment (mschreter) [reply include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.json]: I see :D. Thanks a lot!

File: include/meltpooldg/interface/parameters.hpp [hunk 25]
```diff
@@ -162,18 +165,21 @@ namespace MeltPoolDG
...
       //   during NavierStokesMatrix::begin_densities() and
       //   NavierStokesMatrix::begin_viscosity(). However, we do not actually
       //   use these values, since we fill the density and viscosity 
       //   differently.
-        adaflo_params.params.density_diff         = 1.0;
-        adaflo_params.params.viscosity_diff       = 1.0;
+        adaflo_params.params.density_diff         = 1.0; 
+        adaflo_params.params.viscosity_diff       = 1.0; 
+        adaflo_params.params.density              = flow.density;
+        adaflo_params.params.viscosity            = flow.viscosity;
```
Comment (peterrum) [inline include/meltpooldg/interface/parameters.hpp:182]: We do not need to set these values. We are setting these values manually.

File: include/meltpooldg/interface/parameters.hpp [hunk 26]
```diff
@@ -77,13 +77,16 @@ namespace MeltPoolDG
   template <typename number = double>
   struct FlowData
   {
+    number      density = 0.0;
```
Comment (peterrum) [inline include/meltpooldg/interface/parameters.hpp:80]: I think gravity is missing?

File: include/meltpooldg/interface/simulationbase.hpp [hunk 27]
```diff
@@ -182,7 +186,7 @@ namespace MeltPoolDG
       if( !boundary_conditions_map[operation_name] )
         boundary_conditions_map[operation_name] = std::make_shared<BoundaryConditions<dim>>();
       
-      auto bc = boundary_conditions_map[operation_name]->no_slip_bc;
+      auto& bc = boundary_conditions_map[operation_name]->no_slip_bc;
```
Comment (peterrum) [inline include/meltpooldg/interface/simulationbase.hpp:189]: LOL ;)

Comment (mschreter) [reply include/meltpooldg/interface/simulationbase.hpp:189]: ...

Comment (peterrum) [reply include/meltpooldg/interface/simulationbase.hpp:189]: The test case `spurious_currents` still has artifacts at the boundaries!?

**Comment by mschreter** [comment]
Unfortunately, yes. The velocity field is simply wrong. I don't know why.

Am Fr., 30. Okt. 2020 um 20:19 Uhr schrieb Peter Munch <
notifications@github.com>:

File: include/meltpooldg/simulations/spurious_currents/spurious_currents.hpp [hunk 28]
```diff
@@ -86,7 +88,8 @@ namespace MeltPoolDG
           if constexpr (dim == 2)
             {
               GridGenerator::hyper_cube (*this->triangulation, -2.5, 2.5);
-              this->triangulation->refine_global(6 /*TODO: parameter?*/);
+              this->triangulation->refine_global(this->parameters.base.global_refinements);
```
Comment (peterrum) [inline include/meltpooldg/simulations/spurious_currents/spurious_currents.hpp]: Thanks for implementing my TODOs :P

File: include/meltpooldg/simulations/rising_bubble/rising_bubble.hpp [hunk 29]
```diff
@@ -1,4 +1,6 @@
 #pragma once
+
+#ifdef MELT_POOL_DG_WITH_ADAFLO
```
Comment (peterrum) [inline include/meltpooldg/simulations/rising_bubble/rising_bubble.hpp:3]: Why is this needed? As far as I see it this should not be necessary here (only where the wrapper is defined).

File: include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.hpp [hunk 30]
```diff
@@ -1,4 +1,6 @@
 #pragma once
+
+#ifdef MELT_POOL_DG_WITH_ADAFLO
```
Comment (peterrum) [inline include/meltpooldg/simulations/flow_past_zylinder/flow_past_zylinder.hpp:3]: same here

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 31]
```diff
@@ -56,13 +56,13 @@ namespace MeltPoolDG
 
         // initialize phases and force(?) [TODO] 
         update_phases(level_set_operation.solution_level_set, base_in->parameters);
+
         // accumulate forces: a) gravity force
-        compute_gravity_force(force_rhs, base_in->parameters.base.gravity, true);
+        compute_gravity_force(force_rhs, base_in->parameters.base.gravity);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp]: I think I intended to say: zero out values...

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 32]
```diff
@@ -339,9 +315,11 @@ namespace MeltPoolDG
 
         create_density_dof_vector(density);
         create_viscosity_dof_vector(viscosity);
+
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp]: I am curious what is the goal of these vectors? Is it only for visualization or do we want to fill these vectors and based on these the internal data structures of adaflo are set up? If latter, we might want to add such a setter function directly to the adflo NS interface.

Comment (mschreter) [reply include/meltpooldg/flow/two_phase_flow_problem.hpp]: Atm these functions should only provide output-routines. To this end, I am currently struggling with the `dealii:VectorTools::project` utility based on the matrix-free container to reuse our `get_density function`.

**Comment by mschreter** [comment]
Today's result -- comparison of rising bubble with adaflo (green dots)

anim

File: include/meltpooldg/flow/two_phase_flow_problem.hpp (hunk 32 above)
Comment (peterrum) [reply include/meltpooldg/flow/two_phase_flow_problem.hpp]: I would recommend not to use `dealii:VectorTools::project` but to do projection in a matrix-free loop. I'll dig out tomorrow a code you to set up an appropriate element projection matrix.

Comment (peterrum) [reply include/meltpooldg/flow/two_phase_flow_problem.hpp]: You can set up the element-projection matrix like this (assuming that we have defined 2 1D-FEs - the source one defined in Gauss-Legendre points (for this FE_DGQArbitraryNodes can be used) and the target points int the Gauss-Lobatto points): https://github.com/dealii/dealii/blob/4145598e03ef39c8d8d4e097b93a9c6009353db4/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L1527-L1536

and then project with sum-factorization like this:

https://github.com/dealii/dealii/blob/4145598e03ef39c8d8d4e097b93a9c6009353db4/include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h#L107-L119

Finally, you can simply call `FEEvaluation::set_dof_values`.

Comment (peterrum) [reply include/meltpooldg/flow/two_phase_flow_problem.hpp]: Here is a draft for the discussion offline:
```cpp
#include <deal.II/base/quadrature_lib.h>
#include <deal.II/base/vectorization.h>

#include <deal.II/fe/fe_dgq.h>
#include <deal.II/fe/fe_tools.h>

#include <deal.II/matrix_free/evaluation_kernels.h>
#include <deal.II/matrix_free/fe_evaluation.h>
#include <deal.II/matrix_free/matrix_free.h>

using namespace dealii;

template <int dim = 2, typename Number = double>
void
test()
{
  // some parameters
  const unsigned int fe_degree     = 2;
  const unsigned int n_q_points_1D = fe_degree + 1;

  // create 2 1D FEs:
  FE_DGQArbitraryNodes<1> fe_coarse(QGauss<1>(n_q_points_1D).get_points());
  FE_DGQ<1>               fe_fine(fe_degree);

  // create 1D projection matrix for sum factorization
  FullMatrix<double> matrix(fe_fine.dofs_per_cell, fe_coarse.dofs_per_cell);
  FETools::get_projection_matrix(fe_coarse, fe_fine, matrix);

  AlignedVector<VectorizedArray<Number>> prolongation_matrix_1d(
    fe_fine.dofs_per_cell * fe_coarse.dofs_per_cell);

  for (unsigned int i = 0, k = 0; i < fe_coarse.dofs_per_cell; ++i)
    for (unsigned int j = 0; j < fe_fine.dofs_per_cell; ++j, ++k)
      prolongation_matrix_1d[k] = matrix(j, i);

  // matrix-free stuff
  MatrixFree<dim, Number, VectorizedArray<Number>> matrix_free; // TODO

  FEEvaluation<dim, -1, 0, 1, Number, VectorizedArray<Number>> fe_eval(
    matrix_free);

  for (unsigned int q = 0; q < fe_eval.n_q_points; ++q)
    fe_eval.begin_values()[q] = 0.0; // TODO -> viscosity/density from adaflo

  // perform basis change from quadrature points to support points
  internal::FEEvaluationImplBasisChange<internal::evaluate_general,
                                        internal::EvaluatorQuantity::value,
                                        dim,
                                        0,
                                        0,
                                        VectorizedArray<Number>,
                                        VectorizedArray<Number>>::
    do_forward(1,
               prolongation_matrix_1d,
               fe_eval.begin_values(),
               fe_eval.begin_dof_values(),
               n_q_points_1D, // number of quadrature points
               fe_degree + 1  // number of support points
    );

  // write values back into global vector
  Vector<Number> v; // TODO
  fe_eval.set_dof_values(v);
}

int
main()
{
  test();
}
```
This code is not tested. This would be one option where we want to plot the data inside of adaflo (directly). However, later we might already have the viscosity and density in global vectors which are used to set the values inside of adaflo: in this case the visualization should be straight forward.

File: .github/workflows/tests.yml [hunk 33]
```diff
@@ -26,6 +36,22 @@ jobs:
     - uses: actions/checkout@v2
     - name: setup
       run: |
-         docker pull dealii/dealii:master-focal
-         docker run -t dealii/dealii:master-focal /bin/sh -c "git clone https://github.com/kronbichler/adaflo.git && mkdir adaflo-build && cd adaflo-build && cmake ../adaflo && make -j2 adaflo && git clone https://github.com/${{ github.repository }} && cd MeltPoolDG && git fetch origin ${{ github.ref }} && git checkout FETCH_HEAD && git --no-pager  log | head -n50 && cd .. && mkdir MeltPoolDG-build && cd MeltPoolDG-build && cmake ../MeltPoolDG -D ADAFLO_LIB=/home/dealii/adaflo-build/ -D ADAFLO_INCLUDE=/home/dealii/adaflo/include/ && make -j2 && ctest -V";
+         "git clone https://github.com/MeltPoolDG/adaflo.git    && \
```
Comment (peterrum) [inline .github/workflows/tests.yml:39]:
```suggestion
         command="git clone https://github.com/MeltPoolDG/adaflo.git    && \
```
This should fix the the failing adaflo CI tests. Sorry about this! Could you take a look at the failing tests without adaflo support and indent the code. I will have the current version of the code in the morning and will merge: any further improvements can be added in follow-up PRs!

**Comment by mschreter** [comment]
At the moment we have problem with the following test files:

- simplex support integration (missing quadrature points
- adaflo (the output includes wall times --> I will check tomorrow how wall times can be excluded from output in adaflo)

**Comment by peterrum** [comment]
> At the moment we have problem with the following test files:
simplex support integration (missing quadrature points

I'll fix that (we have changed finally the way we define the number of quadrature points -  see https://github.com/dealii/dealii/pull/11103)

> adaflo (the output includes wall times --> I will check tomorrow how wall times can be excluded from output in adaflo)

I hope there is an input parameter.

What kinda worries me that the adaflo build has a linking error!?

**Comment by peterrum** [comment]
I have changed the clang-format version in the indentation test from 11 to 10 because the test kept on failing.

**Comment by peterrum** [comment]
Locally all tests are passing (with and without adflo enabled).

File: include/meltpooldg/advection_diffusion/advection_diffusion_problem.hpp [hunk 34]
```diff
@@ -146,9 +146,8 @@ namespace MeltPoolDG
         unsigned int quad_idx = 0;
 #ifdef DEAL_II_WITH_SIMPLEX_SUPPORT
         if (base_in->parameters.base.do_simplex)
-          quad_idx = scratch_data->attach_quadrature(Simplex::QGauss<dim>(
-            dim == 2 ? (base_in->parameters.base.n_q_points_1d == 1 ? 3 : 7) :
-                       (base_in->parameters.base.n_q_points_1d == 1 ? 4 : 10)));
+          quad_idx = scratch_data->attach_quadrature(
+            Simplex::QGauss<dim>(base_in->parameters.base.n_q_points_1d));
```
Comment (peterrum) [inline include/meltpooldg/advection_diffusion/advection_diffusion_problem.hpp:150]: This it how it looks like now <3

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 35]
```diff
@@ -32,6 +33,21 @@ namespace MeltPoolDG
...
+     * particularly suited in the context of MatrixFree routines.
+     */
+    template <typename number>
+    VectorizedArray<number>
+    heaviside(const VectorizedArray<number> &in, const number limit = 0.0)
+    {
+      VectorizedArray<number> out;
+      for (unsigned int v = 0; v < VectorizedArray<number>::size(); ++v)
+        out = (in[v] > limit) ? 1 : 0;
+      return out;
```
Comment (peterrum) [inline include/meltpooldg/utilities/utilityfunctions.hpp:48]:
```suggestion
      return compare_and_apply_mask<SIMDComparison::greater_than>  (v, VectorizedArray<double>(limit), 1.0, 0.0);
```

File: include/meltpooldg/simulations/flow_past_cylinder/flow_past_cylinder.hpp [hunk 36]
```diff
@@ -0,0 +1,196 @@
...
+
+#include <iostream>
+// MeltPoolDG
+#include <meltpooldg/interface/simulationbase.hpp>
+
+namespace MeltPoolDG
+{
+  namespace Simulation
+  {
+    namespace FlowPastCylinder
```
Comment (peterrum) [inline include/meltpooldg/simulations/flow_past_cylinder/flow_past_cylinder.hpp:21]: Is this test case now actually simulate flow-past cylinder (single phase flow)? If not, I guess we could re-enable it later.

File: include/meltpooldg/level_set/level_set_operation.hpp [hunk 37]
```diff
@@ -128,7 +134,61 @@ namespace MeltPoolDG
...
+      /*
+       *  getter functions for solution vectors
+       */
+      // @ todo
+
     private:
+      void
+      transform_level_set_to_heaviside()
+      {
+        // @todo
```
Comment (peterrum) [inline include/meltpooldg/level_set/level_set_operation.hpp:189]: Assert?

File: include/meltpooldg/level_set/level_set_operation.hpp [hunk 38]
```diff
@@ -128,7 +134,61 @@ namespace MeltPoolDG
         curvature_operation.solve(advec_diff_operation.solution_advected_field);
       }
 
+      void
+      compute_surface_tension(BlockVectorType &force_rhs,
+                              const double     surface_tension_coefficient,
+                              const bool       add = false)
```
Comment (peterrum) [inline include/meltpooldg/level_set/level_set_operation.hpp:140]: What about 
```suggestion
                              const bool       zero_out = true)
```

File: include/meltpooldg/flow/flow_base.hpp [hunk 39]
```diff
@@ -0,0 +1,38 @@
...
+  namespace Flow
+  {
+    class FlowBase
+    {
+    public:
+      virtual void
+      solve() = 0;
+
+      virtual void
+      get_velocity(LinearAlgebra::distributed::BlockVector<double> &vec) const = 0;
```
Comment (peterrum) [inline include/meltpooldg/flow/flow_base.hpp:19]: I'll add here some comments once I have some time^^

File: include/meltpooldg/flow/two_phase_flow_problem.hpp [hunk 40]
```diff
@@ -0,0 +1,404 @@
...
+              base_in->parameters.flow.surface_tension_coefficient,
+              true /*add to force vector*/);
+
+            //  ... and set forces within the Navier-Stokes solver
+            flow_operation->set_force_rhs(force_rhs);
+            // solver Navier-Stokes problem
+            flow_operation->solve();
+
+            // extract velocity form Navier-Stokes solver ...
+            flow_operation->get_velocity(advection_velocity);
```
Comment (peterrum) [inline include/meltpooldg/flow/two_phase_flow_problem.hpp:81]: I see you swapped the order of LS-NS :+1:

**Review by peterrum — APPROVED** [review]
Looks good to me; I am happy to merge. But don't blame me if I missing something: going throw more than 2000 changed/added locs is a lot and the probility to miss some details is quite high ;)

I don't know why `GitHub CI / build-adaflo` keeps on having linking errors. I have run the tests locally and everything is fine.

File: include/meltpooldg/level_set/level_set_operation.hpp (hunk 37 above)
Comment (mschreter) [reply include/meltpooldg/level_set/level_set_operation.hpp:189]: done.

File: include/meltpooldg/level_set/level_set_operation.hpp (hunk 38 above)
Comment (mschreter) [reply include/meltpooldg/level_set/level_set_operation.hpp:140]: done.

File: include/meltpooldg/utilities/utilityfunctions.hpp [hunk 41]
```diff
@@ -42,10 +42,7 @@ namespace MeltPoolDG
     VectorizedArray<number>
     heaviside(const VectorizedArray<number> &in, const number limit = 0.0)
     {
-      VectorizedArray<number> out;
-      for (unsigned int v = 0; v < VectorizedArray<number>::size(); ++v)
-        out = (in[v] > limit) ? 1 : 0;
-      return out;
+      return compare_and_apply_mask<SIMDComparison::greater_than>  (v, VectorizedArray<double>(limit), 1.0, 0.0);
```
Comment (peterrum) [inline include/meltpooldg/utilities/utilityfunctions.hpp:45]:
```suggestion
      return compare_and_apply_mask<SIMDComparison::greater_than>  (in, VectorizedArray<double>(limit), 1.0, 0.0);
```

Indentation is also off. Feel free to merge once you are happy!

File: include/meltpooldg/interface/parameters.hpp [hunk 42]
```diff
@@ -20,6 +23,7 @@ namespace MeltPoolDG
     int          n_q_points_1d       = -1;
     bool         do_print_parameters = true;
     bool         do_simplex          = false;
+    number       gravity             = 0.981;
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/interface/parameters.hpp:26]: Since its value depends on the unit system (and on the planet ;) ), I will set the default value to zero.

[the remark this answers is not in the document]
Comment (peterrum) [reply include/meltpooldg/interface/parameters.hpp:26]: There should be also a significant difference between the development sites :P
